### PR TITLE
Userland: Move `LibShell` to be in its own directory, introduce `LibShellMinimal` and use it within BuggieBox

### PR DESCRIPTION
**Pull request by supercomputer7**

### CODE REVIEW

File: Userland/Shell/CMakeLists.txt [hunk 1]
```diff
@@ -4,11 +4,21 @@ serenity_component(
...
 if (ENABLE_FUZZERS)
     add_custom_target(Shell) # Stub target to placate serenity_component
+    add_custom_target(ShellMinimal) # Stub target to placate serenity_component
 else()
     set(SOURCES main.cpp)
 
     serenity_bin(Shell)
     target_link_libraries(Shell PRIVATE LibCore LibShell LibFileSystem LibLine LibMain)
+
+    serenity_bin(ShellMinimal)
```
Comment (alimpfard) [inline Userland/Shell/CMakeLists.txt:22]: uhh...but why?
Nothing uses this, and from the user's point of view this and the regular shell are the exact same

File: Userland/Libraries/LibShell/AST.cpp [hunk 2]
```diff
@@ -14,12 +14,15 @@
 #include <AK/StringBuilder.h>
 #include <LibCore/EventLoop.h>
 #include <LibFileSystem/FileSystem.h>
-#include <LibURL/URL.h>
 #include <errno.h>
 #include <fcntl.h>
 #include <signal.h>
 #include <unistd.h>
 
+#ifdef USE_LIBURL
```
Comment (alimpfard) [inline Userland/Libraries/LibShell/AST.cpp:22]: No way, nu-uh.
Please come up with a better solution, or simply accept the overhead of including those libraries.

File: Userland/Shell/Tests/test-commons.inc [hunk 3]
```diff
@@ -1,11 +0,0 @@
...
-    # The AST ends up at stderr, and CI checks for 'fail' (uppercase)
-    # Trick it into not failing the test.
-    echo "FA""IL:" $*
-    exit 1
-}
-
-pass() {
-    echo "PA""SS"
-    exit 0
-}
```
Comment (alimpfard) [inline Userland/Shell/Tests/test-commons.inc:11]: why is this removed?

**Review by alimpfard — COMMENTED** [review]
*all* of the libraries in the system equate to about 45M, taking out just the leaf web-only stuff, you're left with ~15M.
What size are you targeting exactly?

File: Userland/Shell/Tests/test-commons.inc (hunk 3 above)
Comment (supercomputer7) [reply Userland/Shell/Tests/test-commons.inc:11]: Not sure why... I moved it to the new directory but the diff says otherwise :man_shrugging:

File: Tests/LibShell/Tests/test-commons.inc [hunk 4]
```diff
@@ -0,0 +1,14 @@
+fail()
+{
+#The AST ends up at stderr, and CI checks for 'fail'(uppercase)
+#Trick it into not failing the test.
+    echo "FA"
+         "IL:" $*
```
Comment (alimpfard) [inline Tests/LibShell/Tests/test-commons.inc:6]: Ah, that's because you somehow reformatted this file.
Please don't, this isn't even correct :)

Comment (supercomputer7) [reply Tests/LibShell/Tests/test-commons.inc]: Done :)

**Comment by supercomputer7** [comment]
> _all_ of the libraries in the system equate to about 45M, taking out just the leaf web-only stuff, you're left with ~15M. What size are you targeting exactly?

I'd like to trim every byte that I can before I can merge everything into the `BuggieBox` program. This is only the first PR in the topic, because I'd like to examine trimming down `LibRegex` and `LibUnicode` as well for `BuggieBox`.

File: Userland/Libraries/LibCore/CMakeLists.txt [hunk 5]
```diff
@@ -24,21 +24,31 @@ if (SERENITYOS)
...
     DateTime.cpp
     ElapsedTimer.cpp
     Event.cpp
     EventLoop.cpp
     EventLoopImplementation.cpp
     EventLoopImplementationUnix.cpp
     EventReceiver.cpp
+    ThreadEventQueue.cpp
+    Notifier.cpp
+    Timer.cpp
```
Comment (timschumi) [inline Userland/Libraries/LibCore/CMakeLists.txt:39]: Please keep this sorted alphabetically.

**Review by timschumi — CHANGES_REQUESTED** [review]

**Comment by timschumi** [comment]
@supercomputer7 Do you have some statistics on hand on how much space these changes save?

**Comment by supercomputer7** [comment]
> @supercomputer7 Do you have some statistics on hand on how much space these changes save?

I briefly did a check and the results are:
```
[liav@liavpc Root]$ ls -la usr/lib | grep libshell | grep -v "libshellcomprehension.so"
lrwxrwxrwx 1 liav liav        27 Aug 10 20:17 libshellminimal.so -> libshellminimal.so.serenity
-rwxr-xr-x 1 liav liav   5756552 Aug 14 17:22 libshellminimal.so.serenity
lrwxrwxrwx 1 liav liav        20 Aug 10 20:17 libshell.so -> libshell.so.serenity
-rwxr-xr-x 1 liav liav   5870912 Sep  3 12:14 libshell.so.serenity
```

So it saves about 100 kilobytes, not really that bad :)
It probably saves a bit more, because we don't need to load `LibURL` or `LibSyntax` when using the reduced version so this could be useful for `BuggieBox` later on

**Review by timschumi — APPROVED** [review]
LGTM, rerunning the failed CI job. Maybe Ali has some more comments in the meantime?

**Comment by supercomputer7** [comment]
> LGTM, rerunning the failed CI job. Maybe Ali has some more comments in the meantime?

Since Ali hasn't responded, what do we do with this pull request? are we settled that I will rebase it and then get it merged? or do you have any suggestion on this change?

**Comment by timschumi** [comment]
> > LGTM, rerunning the failed CI job. Maybe Ali has some more comments in the meantime?
> 
> Since Ali hasn't responded, what do we do with this pull request? are we settled that I will rebase it and then get it merged? or do you have any suggestion on this change?

I kind of forgot that there was a pending CI run, no more comments from my side. Let's get it rebased one last time.
